src/workerd/server/server.c++

```diff
@@ -5115,6 +5126,145 @@ kj::Promise<void> Server::listenHttp(kj::Own<kj::ConnectionReceiver> listener,
   co_return co_await obj->run();
 }
 
+// =======================================================================================
+// Debug port for exposing all services via RPC
+
+class Server::DebugPortListener final: public kj::Refcounted, private kj::TaskSet::ErrorHandler {
+ public:
+  DebugPortListener(Server& owner,
+      kj::Own<kj::ConnectionReceiver> listener,
+      capnp::HttpOverCapnpFactory& httpOverCapnpFactory)
+      : owner(owner),
+        listener(kj::mv(listener)),
+        httpOverCapnpFactory(httpOverCapnpFactory),
+        tasks(*this) {}
+
+  kj::Promise<void> run() {
+    for (;;) {
+      auto conn = co_await listener->accept();
+      tasks.add(acceptLoop(kj::mv(conn)));
+    }
+  }
+
+ private:
+  Server& owner;
+  kj::Own<kj::ConnectionReceiver> listener;
+  capnp::HttpOverCapnpFactory& httpOverCapnpFactory;
+  kj::TaskSet tasks;
+
+  kj::Promise<void> acceptLoop(kj::Own<kj::AsyncIoStream> conn) {
+    auto server = kj::heap<capnp::TwoPartyServer>(
+        kj::heap<WorkerdDebugPortImpl>(&owner, httpOverCapnpFactory));
+    auto promise = server->accept(*conn);
+    co_return co_await promise.attach(kj::mv(conn), kj::mv(server));
+  }
+
+  void taskFailed(kj::Exception&& exception) override {
+    KJ_LOG(ERROR, exception);
+  }
+
+  class WorkerdDebugPortImpl final: public rpc::WorkerdDebugPort::Server {
+   public:
+    WorkerdDebugPortImpl(
+        workerd::server::Server* srvPtr, capnp::HttpOverCapnpFactory& httpOverCapnpFactory)
+        : srv(*srvPtr),
+          httpOverCapnpFactory(httpOverCapnpFactory) {}
+
+    kj::Promise<void> getEntrypoint(GetEntrypointContext context) override {
+      auto params = context.getParams();
+      auto serviceName = params.getService();
+      auto entrypointName = params.getEntrypoint();
+      auto propsReader = params.getProps();
+
+      // Look up the service
+      auto& serviceEntry =
+          KJ_ASSERT_NONNULL(srv.services.find(serviceName), "Service not found", serviceName);
+      auto service = serviceEntry->service();
+
+      // Convert props from Frankenvalue
+      Frankenvalue props = Frankenvalue::fromCapnp(propsReader);
```

Comment:
Probably should check `params.hasProps()` and skip calling `fromCapnp()` on default props. (A default-initialized Frankenvalue is all you want.)